File: Swift/LeetCode/LeetCode/Combinations.swift
//
//  Combinations.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 05/01/2018.
//  Copyright © 2018 lizhaotailang. All rights reserved.
//
//  Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.
//
//  For example,
//  If n = 4 and k = 2, a solution is:
//
//  [
//  [2,4],
//  [3,4],
//  [2,3],
//  [1,2],
//  [1,3],
//  [1,4],
//  ]
//
//  See [CombinationsTests](./LeetCodeTests/CombinationsTests.swift) for test cases.
//

import Foundation

class Combinations {
    
//    Recursive Solution.
//    Accepted.
//    func combine(_ n: Int, _ k: Int) -> [[Int]] {
//        var results = Array<Array<Int>>.init()
//        if n == 0 || k == 0 || k > n {
//            return results
//        }
//
//        if k == 1 {
//            var i = 1
//            while i <= n {
//                results.append([i])
//
//                i += 1
//            }
//
//            return results
//        }
//
//        for list in combine(n, k - 1) {
//            var i = list[list.count - 1]
//            while i < n {
//                var tmp = Array<Int>.init()
//                tmp.append(contentsOf: list)
//                tmp.append(i + 1)
//                results.append(tmp)
//
//                i += 1
//            }
//        }
//
//        return results
//    }
    
//    Iterative Solution.
//    Accepted.
    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        var results = Array<Array<Int>>.init()
        if n == 0 || k == 0 || k > n {
            return results
        }

        var i = 1
        while i <= n + 1 - k {
            results.append([i])

            i += 1
        }

        i = 2
        while i <= k {
            var tmp = Array<Array<Int>>.init()
            for list in results {
                var m = list[list.count - 1] + 1
                while m <= n - (k - i) {
                    var newList = Array<Int>.init()
                    newList.append(contentsOf: list)
                    newList.append(m)
                    tmp.append(newList)

                    m += 1
                }
            }
            results = tmp

            i += 1
        }

        return results
    }
    
}

File: JavaScript/src/Combinations.js
/**
 * Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.
 *
 * For example,
 * If n = 4 and k = 2, a solution is:
 *
 * [
 * [2,4],
 * [3,4],
 * [2,3],
 * [1,2],
 * [1,3],
 * [1,4],
 * ]
 */

/**
 * Iterative solution.
 *
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
let combine = function (n, k) {
    let results = [];
    if (n === 0 || k === 0 || k > n) {
        return results;
    }

    for (let i = 1; i <= n + 1 - k; i++) {
        results.push([i]);
    }

    for (let i = 2; i <= k; i++) {
        let tmp = [];
        results.forEach(function (list) {
            for (let m = list[list.length - 1] + 1; m <= n - (k - i); m++) {
                let newList = [];
                list.forEach(function (element) {
                    newList.push(element);
                });
                newList.push(m);
                tmp.push(newList);
            }
        });
        results = tmp;
    }
    return results;
};

/**
 * Recursive solution.
 * Time Limit Exceeded.
 *
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
/*let combine = function (n, k) {
    let results = [];
    if (n === 0 || k === 0 || k > n) {
        return results;
    }
    if (k === 1) {
        for (let i = 1; i <= n; i++) {
            let list = [];
            list.push(i);
            results.push(list);
        }
        return results;
    }
    combine(n, k - 1).forEach(function (list) {
        for (let i = list[list.length - 1]; i < n; i++) {
            let tmp = [];
            list.forEach(function (a) {
                tmp.push(a);
            });
            tmp.push(i + 1);
            results.push(tmp);
        }
    });
    return results;
};*/

if (combine(2, 0).length === 0) {
    console.log("pass")
} else {
    console.error("failed")
}

let list0 = combine(2, 1);
let tmp = new Set([[1], [2]]);
if (list0.length === 2) {
    console.log("pass")
} else {
    console.error("failed")
}
if (new Set(list0).toString() === tmp.toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

tmp.clear();
let list1 = combine(4, 2);
tmp.add([[2, 4], [3, 4], [2, 3], [1, 2], [1, 3], [1, 4]]);
if (list1.length === 6) {
    console.log("pass")
} else {
    console.error("failed")
}
if (new Set(list1).toString() === tmp.toString()) {
    console.log("pass")
} else {
    console.error("failed")
}

tmp.clear();
let list2 = combine(4, 3);
tmp.add([[1, 2, 3], [1, 2, 4], [1, 3, 4], [2, 3, 4]]);
if (list2.length === 4) {
    console.log("pass")
} else {
    console.error("failed")
}
if (new Set(list2).toString() === tmp.toString()) {
    console.log("pass")
} else {
    console.error("failed")
}
